py/htm/encoders/eye.py

```diff
@@ -0,0 +1,627 @@
+        self.position    = [random.uniform(roi_radius, dim - roi_radius)
+                                 for dim in self.image.shape[:2]]
+
+    def _crop_roi(self):
+        """
+        Crop to Region Of Interest (ROI) which contains the whole field of view.
+        Adds a black circular boarder to mask out areas which the eye can't see.
+
+        Note: size of the ROI is (eye.output_diameter * eye.resolution_factor).
+        Note: the circular boarder is actually a bit too far out, playing with
+          eye.fovea_scale can hide areas which this ROI image will show.
+
+        Arguments: eye.scale, eye.position, eye.image
+
+        Returns RGB image (diameter * diameter, but effectively cropped to an 
+          inner circle - FOV). 
+
+        See also, @see make_roi_pretty()
+        """
+        assert(self.image is not None)
+
+        r     = int(round(self.scale * self.retina_diameter / 2))
+        x, y  = self.position
+        x     = int(round(x))
+        y     = int(round(y))
+        x_max, y_max, color_depth = self.image.shape
+
+        # Find the boundary of the ROI and slice out the image.
+        x_low  = max(0, x-r)
+        x_high = min(x_max, x+r)
+        y_low  = max(0, y-r)
+        y_high = min(y_max, y+r)
+        image_slice = self.image[x_low : x_high, y_low : y_high]
+
+        # Make the ROI and insert the image into it.
+        roi = np.zeros((2*r, 2*r, 3,), dtype=np.uint8)
+        if x-r < 0:
+            x_offset = abs(x-r)
+        else:
+            x_offset = 0
+        if y-r < 0:
+            y_offset = abs(y-r)
+        else:
+            y_offset = 0
+        x_shape, y_shape, color_depth = image_slice.shape
+        roi[x_offset:x_offset+x_shape, y_offset:y_offset+y_shape] = image_slice
+
+        # Rescale the ROI to remove the scaling effect.
+        roi.resize( (self.retina_diameter, self.retina_diameter, 3))
+
+        # Mask out areas the eye can't see by drawing a circle boarder.
+        # this represents the "shape" of the sensor/eye (comment out to leave rectangural)
+        center = int(roi.shape[0] / 2)
+        circle_mask = np.zeros(roi.shape, dtype=np.uint8)
+        cv2.circle(circle_mask, (center, center), center, thickness = -1, color=(255,255,255))
+        roi = np.minimum(roi, circle_mask)
+
+        return roi
+
+
+    def compute(self, position=None, rotation=None, scale=None):
+        """
+        Arguments position, rotation, scale: optional, if not None, the self.xxx is overriden
+          with the provided value.
+        Returns tuple (SDR parvo, SDR magno) 
+        """
+        # set position
+        if position is not None:
+          self.position = position
+        if rotation is not None:
+          self.orientation=rotation
+        if scale is not None:
+          self.scale=scale
+
+        # apply field of view (FOV), rotation
+        self.roi = self._crop_roi()
+        self.roi = self.rotate_(self.roi, self.orientation)
```

Comment:
apply rotation to the image itself, instead of separately to output, visualizations, etc